komodo/build.py

```diff
@@ -77,58 +79,72 @@ def cmake(pkg, ver, path, prefix, builddir,
         print(shell('make DESTDIR={} install'.format(fakeroot)))
 
-def sh(pkg, ver, pkgpath, prefix, makefile, *args, **kwargs):
-    if os.path.isfile(makefile):
-        makefile = os.path.abspath(makefile)
+
+def sh(pkg, ver, pkgpath, data, prefix, makefile, *args, **kwargs):
+    makefile = data.get(makefile)
 
     with pushd(pkgpath):
-        makefile = os.path.abspath(makefile)
         cmd = ['bash {} --prefix {}'.format(makefile, prefix),
                '--fakeroot {}'.format(kwargs['fakeroot']),
                '--python {}/bin/python'.format(prefix)]
         if 'jobs' in kwargs:
             cmd.append('--jobs {}'.format(kwargs['jobs']))
         if 'cmake' in kwargs:
             cmd.append('--cmake {}'.format(kwargs['cmake']))
+        cmd.append('--pythonpath {}'.format(kwargs['pythonpath']))
+        cmd.append('--path {}'.format(kwargs['binpath']))
+        cmd.append('--pip {}'.format(kwargs['pip']))
+        cmd.append('--virtualenv {}'.format(kwargs['virtualenv']))
         cmd.append(kwargs.get('makeopts'))
 
         print('Installing {} ({}) from sh'.format(pkg, ver))
         shell(cmd)
 
-def rsync(pkg, ver, pkgpath, prefix, *args, **kwargs):
+
+def rsync(pkg, ver, pkgpath, data, prefix, *args, **kwargs):
     print('Installing {} ({}) with rsync'.format(pkg, ver))
     # assume a root-like layout in the pkgpath dir, and just copy it
     shell(['rsync -am', kwargs.get('makeopts'), '{}/'.format(pkgpath), kwargs['fakeroot'] + prefix])
 
 
-def pip_install(pkg, ver, pkgpath, prefix, dlprefix, pip='pip', *args, **kwargs):
+def pip_install(pkg, ver, pkgpath, data, prefix, dlprefix, pip='pip', *args, **kwargs):
     cmd = [pip,
            'install {}=={}'.format(pkg, komodo.strip_version(ver)),
            '--root {}'.format(kwargs['fakeroot']),
            '--prefix {}'.format(prefix),
-           '--force-reinstall',
            '--no-index',
            '--no-deps',
+           '--ignore-installed',
            '--find-links {}'.format(dlprefix),
-           kwargs.get('makeopts', '')]
+    ]
+    cmd.append(kwargs.get('makeopts', ''))
 
     print('Installing {} ({}) from pip'.format(pkg, ver))
     shell(cmd)
 
+
+def noop(pkg, ver, *args, **kwargs):
+    print('Doing nothing for noop package {} ({})'.format(pkg, ver))
+    pass
+
+
 def pypaths(prefix, version):
     if version is None: return ''
     pyver = 'python' + '.'.join(version.split('.')[:-1])
     return ':'.join([ '{0}/lib/{1}'.format(prefix, pyver),
                       '{0}/lib/{1}/site-packages'.format(prefix, pyver),
                       '{0}/lib64/{1}/site-packages'.format(prefix, pyver) ])
 
-def make(pkgfile, repofile, prefix = None,
-                            dlprefix = None,
-                            builddir = None,
-                            jobs = 1,
-                            cmk = 'cmake',
-                            pip = 'pip',
-                            fakeroot = '.'):
+
+def make(pkgfile, repofile, data,
```

Comment:
I think we should create a single `Configuration` object that has most of this information (function args) rather than passing a million different ones separately.

We don't have to do it in this PR, but I'd like to join in writing it. Wasn't entirely sure how to do this pythonically so I stopped trying to implement it myself.

Reply:
Yeah, let's do that pairwise. I'm not sure either.